internal/server/apimigrator.go

```diff
@@ -0,0 +1,264 @@
+package server
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"reflect"
+
+	"github.com/gin-gonic/gin"
+)
+
+type apiMigration struct {
+	method          string
+	path            string
+	version         string
+	redirect        string
+	requestRewrite  func(c *gin.Context)
+	responseRewrite func(c *gin.Context)
+}
+
+func addRedirect(a *API, method, path, newPath, version string) {
+	a.migrations = append(a.migrations, apiMigration{
+		method:   method,
+		path:     path,
+		version:  version,
+		redirect: newPath,
+	})
+}
+
+func addRequestRewrite[oldReq any, newReq any](a *API, method, path, version string, f func(oldReq) newReq) {
+	a.migrations = append(a.migrations, apiMigration{
+		method:  method,
+		path:    path,
+		version: version,
+		requestRewrite: func(c *gin.Context) {
+			reqVer := NewVersion(c.Request.Header.Get("VERSION"))
+			if reqVer.GreaterThanStr(version) {
+				c.Next()
+				return
+			}
+
+			oldReqObj := new(oldReq)
+
+			err := bind(c, oldReqObj)
+			if err != nil {
+				sendAPIError(c, err)
```

Comment:
We should probably add a trace to this error or it will cause a lot of pain to troubleshoot